node/src/components/chain_synchronizer.rs

```diff
@@ -374,97 +417,189 @@ impl ChainSynchronizer {
             result,
         })
     }
 
     fn handle_execute_immediate_switch_block_result(
         &mut self,
-        effect_builder: EffectBuilder<JoinerEvent>,
-        maybe_upgrade_block_header: Option<BlockHeader>,
+        effect_builder: EffectBuilder<REv>,
+        maybe_switch_block_header_before_upgrade: Option<BlockHeader>,
+        is_emergency_upgrade: bool,
         result: Result<BlockAndExecutionEffects, BlockExecutionError>,
     ) -> Effects<Event> {
-        let block_and_execution_effects = match result {
+        let immediate_switch_block_and_exec_effects = match result {
             Ok(block_and_execution_effects) => block_and_execution_effects,
             Err(error) => {
                 error!(%error, "failed to execute block");
                 return fatal!(effect_builder, "{}", error).ignore();
             }
         };
 
-        // If the upgrade block is `None`, this is a genesis switch block, so we can use the
-        // validator set from that.
-        let maybe_era_end = maybe_upgrade_block_header
+        // If the switch block before the immediate switch block is `None`, we use the
+        // `next_era_validators` of the immediate switch block to sign it.  This is the case at
+        // genesis and for an emergency upgrade.
+        let maybe_era_end = maybe_switch_block_header_before_upgrade
             .as_ref()
-            .unwrap_or_else(|| block_and_execution_effects.block.header())
+            .unwrap_or_else(|| immediate_switch_block_and_exec_effects.block.header())
             .era_end();
+
         let validators_to_sign_immediate_switch_block = match maybe_era_end {
             Some(era_end) => era_end
                 .next_era_validator_weights()
                 .keys()
                 .cloned()
                 .collect(),
             None => {
-                error!(
-                    ?maybe_upgrade_block_header,
-                    "upgrade/genesis block is not a switch block"
-                );
+                error!("upgrade/genesis switch block missing era end");
                 return fatal!(
                     effect_builder,
-                    "upgrade/genesis block is not a switch block"
+                    "upgrade/genesis switch block missing era end"
                 )
                 .ignore();
             }
         };
+
+        // For an emergency upgrade, we always execute/commit locally rather than syncing over it.
+        // This means we should try fast syncing again if this was an emergency upgrade.
+        if is_emergency_upgrade {
+            return operations::run_fast_sync_task(
+                effect_builder,
+                self.config.clone(),
+                self.metrics.clone(),
+            )
+            .event(|result| Event::FastSyncAfterEmergencyUpgradeResult {
+                immediate_switch_block_and_exec_effects,
+                validators_to_sign_immediate_switch_block,
+                result,
+            });
```

Comment:
And here: This seems to be the only place where the event is created, and we already have all the data we need to call `handle_fast_sync_after_emergency_result`, so we could remove the event?

Reply:
Same as above.